### PR TITLE
UCT/IB/TEST: Use memory registration cache in IB transports.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1346/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/737/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1347/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/738/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/739/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1348/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1350/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/741/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@alex-mikheev @miked-mellanox pls review

**Comment by MattBBaker** [comment]
autobot build

**Comment by shamisp** [comment]
This is very substantial change. @MattBBaker you would have to update ugni. Do we have to update "mm' transports ?

**Comment by yosefe** [comment]
@shamisp this change affects only IB, not any other transports. It makes IB cache the registrations.

File: src/uct/ib/base/ib_device.c [hunk 1]
```diff
@@ -308,16 +150,16 @@ static ucs_status_t uct_ib_device_create(struct ibv_device *ibv_device,
...
         ucs_error("Failed to open %s: %m", ibv_get_device_name(ibv_device));
         status = UCS_ERR_IO_ERROR;
         goto err;
     }
 
     /* Read device properties */
-    IBV_EXP_DEVICE_ATTR_SET_COMP_MASK(&dev_attr);
-    ret = ibv_exp_query_device(ibv_context, &dev_attr);
+    IBV_EXP_DEVICE_ATTR_SET_COMP_MASK(&dev->dev_attr);
+    ret = ibv_exp_query_device(dev->ibv_context, &dev->dev_attr);
```
Comment (miked-mellanox) [inline src/uct/ib/base/ib_device.c]: why exp API is used and upstream one?

Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: if posible, need to use exp api to detect capabilities such as odp, dc, contig pages...

**Comment by yosefe** [comment]
@MattBBaker can you post the failures?

**Comment by MattBBaker** [comment]
@yosefe I'll pull those up

@shamisp It looks like this particular change is IB specific. I'll probably implement something similar for ugni when I add reg cache to that.

**Comment by shamisp** [comment]
@yosefe we will want to enable similar functionality for ugni and potentially shared memory. We have to discuss how the interface maps to the rest transports

**Comment by MattBBaker** [comment]
It was a build failure,
uct/test_pd.cc:11:28: error: ucs/gtest/test.h: No such file or directory
uct/test_pd.cc:13: error: ‘testing’ has not been declared
uct/test_pd.cc:13: error: expected ‘{’ before ‘TestWithParam’
uct/test_pd.cc:13: error: expected initializer before ‘<’ token
make: *** [uct/gtest-test_pd.o] Error 1

I need to poke autobot more tough, it is still black magic to me.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/748/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1358/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/749/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_pd.c [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+                            (uct_mem_h *)&region->mr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    uct_ib_rcfache_region_set_stub_mr(region);
+    return UCS_OK;
+}
+
+static void uct_ib_rcache_mem_dereg_cb(void *context, ucs_rcache_t *rcache,
```
Comment (shamisp) [inline src/uct/ib/base/ib_pd.c]: The de-registration may result in an error. Would it make sense to propagate the error to the upper layer ?

File: src/uct/ib/base/ib_pd.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+#define UCT_IB_PD_H_
+
+#include "ib_device.h"
+
+#include <uct/base/uct_pd.h>
+#include <ucs/stats/stats.h>
+#include <ucs/sys/rcache.h>
+
+
+enum {
```
Comment (shamisp) [inline src/uct/ib/base/ib_pd.h]: please add description for the enum

File: test/gtest/uct/test_pd.cc [hunk 4]
```diff
@@ -92,4 +171,51 @@ UCS_TEST_P(test_pd, reg) {
...
+        free(ptr);
+    }
+}
+
+
+#define UCT_PD_INSTANTIATE_TEST_CASE(_test_case) \
+    UCS_PP_FOREACH(_UCT_PD_INSTANTIATE_TEST_CASE, _test_case, \
+                   knem, \
+                   cma, \
+                   posiz, \
```
Comment (shamisp) [inline test/gtest/uct/test_pd.cc]: typo ? posix ?

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/760/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_pd.c (hunk 2 above)
Comment (yosefe) [inline src/uct/ib/base/ib_pd.c]: the current UCT API does not return error (our general approach is to not return error from cleanup functions / destructors)

Comment (shamisp) [inline src/uct/ib/base/ib_pd.c]: ok

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1371/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/761/ for details (Mellanox internal link).

**Comment by miked-mellanox** [comment]
:+1:

**Comment by shamisp** [comment]
:+1:

**Pull request merged**